[Tree Borrows] Full Tree Borrows Support (#221)
minirust/minirust: spec/mem/tree_borrows/state_machine.md

                Permission::Disabled => throw_ub!("Tree Borrows: reading from the child of a pointer with Disabled permission"),
            }
        )
    }

    fn child_write(self, actively_protected: bool) -> Result<Permission> {
        match self {
            Permission::Reserved { conflicted: true } if actively_protected =>
                throw_ub!("Tree Borrows: writing to the child of a pointer with Conflicted Reserved permission"),
            Permission::Reserved { .. } | Permission::Active => ret(Permission::Active),
            Permission::Frozen => throw_ub!("Tree Borrows: writing to the child of a pointer with Frozen permission"),
            Permission::Disabled => throw_ub!("Tree Borrows: writing to the child of a pointer with Disabled permission"),
        }
    }

    fn foreign_read(self, actively_protected: bool) -> Result<Permission> {
        match self {
            Permission::Reserved { .. } if actively_protected => ret(Permission::Reserved { conflicted: true }),
            Permission::Reserved { conflicted } => ret(Permission::Reserved { conflicted }),
            Permission::Active if actively_protected =>

>                 throw_ub!("Tree Borrows: reading from the foreign of an actively protected pointer with Active permission"),

Reviewer: (this is still open)

    fn child_read(self) -> Result<Permission> {
        ret(
            match self {
                Permission::Disabled => throw_ub!("Tree Borrows: reading from the child of a pointer with Disabled permission"),
                _ => self,
            }
        )
    }

    fn child_write(self, protected: Protected) -> Result<Permission> {
        match self {

>             Permission::Reserved { conflicted: true } if protected.yes() =>

Reviewer: Why do some transitions check `protected.yes()` and others `protected.active(accessed)`?

Author: A foreign read should make a non-accessed Reserved becomes `Reserved { conflicted: true }`.

Author: Explicitly splitting these two states is somewhat clear to me, what do you think?

Reviewer 2: > Why do some transitions check `protected.yes()` and others `protected.active(accessed)`?

For reasons outlined [here](https://github.com/minirust/minirust/pull/221#issuecomment-2276884520).

Reviewer 2: Miri achieves this by doing this here: https://github.com/rust-lang/miri/blob/master/src/borrow_tracker/tree_borrows/tree.rs#L111

Reviewer: Yeah I think I'd prefer if we did what Miri does, and have a single "transitions to disabled on initialized protected locations are UB" check.

                throw_ub!("Tree Borrows: writing to the child of a protected pointer with Conflicted Reserved permission"),
            Permission::Frozen => throw_ub!("Tree Borrows: writing to the child of a pointer with Frozen permission"),
            Permission::Disabled => throw_ub!("Tree Borrows: writing to the child of a pointer with Disabled permission"),
            _ => ret(Permission::Active),
        }
    }

>     fn foreign_read(self, protected: Protected, accessed: Accessed) -> Result<Permission> {

Reviewer: In Miri, all transitions are given is a `protected: bool`. I think we should do the same here, let's have the smallest possible amount of degrees of freedom.

Author: Same reason as the above reply.

        match self {
            Permission::Active if protected.active(accessed) =>
                throw_ub!("Tree Borrows: reading from the foreign of an actively protected pointer with Active permission"),
            Permission::Reserved { .. } if protected.yes() => ret(Permission::Reserved { conflicted: true }),
            Permission::Active => ret(Permission::Frozen),
            _ => ret(self),
        }
    }

    fn foreign_write(self, protected: Protected, accessed: Accessed) -> Result<Permission> {

>         if !protected.active(accessed) {

Reviewer: Why does this function look so different from the others?

            Permission::Active => throw_ub!("Tree Borrows: writing to the foreign of an actively protected pointer with Active permission"),
            Permission::Frozen => throw_ub!("Tree Borrows: writing to the foreign of an actively protected pointer with Frozen permission"),
            Permission::Disabled => panic!("Permission::foreign_write: Impossible Combination (Accessed + Protected + Disabled)"),
            Permission::ReservedIM => panic!("Permission::foreign_write: Impossible Combination (Accessed + Protected + ReservedIM)"),
        }
    }

    fn transition(
        self,
        access_kind: AccessKind,
        node_relation: NodeRelation,
        protected: Protected,
        accessed: Accessed,
    ) -> Result<Permission> {
        match (node_relation, access_kind) {
            (NodeRelation::Foreign, AccessKind::Write) => self.foreign_write(protected, accessed),
            (NodeRelation::Foreign, AccessKind::Read) => self.foreign_read(protected, accessed),
            (NodeRelation::Child, AccessKind::Write) => self.child_write(protected),
            (NodeRelation::Child, AccessKind::Read) => self.child_read(),
        }
    }

>     fn default(mutbl: Mutability, pointee: PointeeInfo, protected: Protected) -> Result<Permission> {

Reviewer: This doesn't need `result`.

Also please move this to the top of the impl block.

        match mutbl {
            Mutability::Mutable if !pointee.freeze && protected.no() => ret(Permission::ReservedIM),

>             Mutability::Mutable => ret(Permission::Reserved { conflicted: false }),

Reviewer: This needs a *detailed* comment explaining why we use `Reserved` for protected interior mutable references.

    fn default(mutbl: Mutability, pointee: PointeeInfo, protected: Protected) -> Permission {
        match mutbl {
            Mutability::Mutable if !pointee.freeze && protected.no() => Permission::ReservedIM,
            // We only use `ReservedIM` for *unprotected* mutable references with interior mutability.
            // If the reference is protected, we ignore the interior mutability.
            // An example for why "Protected + Interior Mutability" is undesirable

>             // can be found in tooling/minimize/tests/ub/tree_borrows/protector/reservedim_spurious_write.rs.

Reviewer: This comment should be one line up, above the `if protected.no()`

        )
    }

    fn child_write(self, protected: bool) -> Result<Permission> {
        match self {
            Permission::Reserved { conflicted: true } if protected =>
                throw_ub!("Tree Borrows: writing to the child of a protected pointer with Conflicted Reserved permission"),
            Permission::Frozen => throw_ub!("Tree Borrows: writing to the child of a pointer with Frozen permission"),
            Permission::Disabled => throw_ub!("Tree Borrows: writing to the child of a pointer with Disabled permission"),
            _ => ret(Permission::Active),
        }
    }

    fn foreign_read(self, protected: bool) -> Result<Permission> {
        match self {
            Permission::Active if protected  => ret(Permission::Disabled),
            Permission::Reserved { .. } if protected => ret(Permission::Reserved { conflicted: true }),
            Permission::Active => ret(Permission::Frozen),
            // All other states are kept unchanged.
            perm => ret(perm),
        }
    }

    fn foreign_write(self, protected: bool) -> Result<Permission> {
        match self {

>             Permission::ReservedIM if !protected => ret(Permission::ReservedIM),

Reviewer 2: The `!protected` is unnecessary. It's still there because it being there makes the theorem that nothing protected survives foreign writes much simpler to prove (it holds by definition)

Reviewer: Please add a comment explaining this.

}

impl LocationState {
    /// Create a location state list:
    /// all locations share the given permission.
    fn new_list(permission: Permission, len: Size) -> List<LocationState> {
        let mut location_states = List::new();
        for _ in Int::ZERO..len.bytes() {
            location_states.push(LocationState {
                accessed: Accessed::No,
                permission,
            });
        }

        location_states
    }

    fn transition(
        &mut self,
        access_kind: AccessKind,
        node_relation: NodeRelation,
        protected: bool,
    ) -> Result {
        let old_perm = self.permission;
        self.permission = old_perm.transition(access_kind, node_relation, protected)?;
        self.accessed = self.accessed.transition(node_relation);

        // Protected nodes may never transition to "Disabled, Accessed". That is UB.
        if self.permission == Permission::Disabled && protected && self.accessed == Accessed::Yes {

>             match old_perm {

Reviewer: ```suggestion
            // This is UB, make sure to show a somewhat specific error.
            match old_perm {
```

        &mut self,
        access_kind: AccessKind,
        node_relation: NodeRelation,
        protected: Protected,
    ) -> Result {
        self.permission = self.permission.transition(access_kind, node_relation, protected, self.accessed)?;

>         self.accessed = self.accessed.transition(node_relation);

Reviewer: Protected nodes cannot get disabled. So what I would expect is that there is a check here: if `permission` is now `Disabled` and `protected.yes()`, then we cause UB. Isn't that the entire point of protectors? How can things work when that is not done...?

Author: In this implementation, we cause UB when the permission **becomes** `Disabled`. This is slightly different from Miri, sorry for confusing

Reviewer 2: Yes, it's implemented differently to miri but achieves the same effect. I also think it is somewhat more clean.

Reviewer: Why do you think this is more clean? I find the Miri version more clean -- a single general principle, uniformly enforced, is better than having to copy the same logic in several places.

Author: I also really like the Miri version :)

But when I was implementing this, I thought that this version could explicitly show all the state transitions, directly corresponding to the state machine.

Reviewer: To me, the state machine is what Miri has, and the rule that "when protected transitions to disabled, that's UB" is implemented outside the state machine.

Author: > the rule that "when protected transitions to disabled, that's UB" is implemented outside the state machine.

This makes sense to me. Fixed it :)

        ret(())
    }

    /// Create a location state list:
    /// all locations share the given permission.

>     fn list(permission: Permission, len: Size) -> List<LocationState> {

Reviewer: Maybe call this `new_list` to make it more clear that this is a constructor?

Also please move it to the top of the impl block.

        &mut self,
        access_kind: AccessKind,
        node_relation: NodeRelation,
        protected: bool,
    ) -> Result {
        let permission = self.permission.transition(access_kind, node_relation, protected)?;
        if permission == Permission::Disabled && protected && self.accessed == Accessed::Yes {
            match self.permission {
                Permission::Disabled => panic!("Impossible state combination: Accessed + Protected + Disabled"),
                Permission::ReservedIM => panic!("Impossible state combination: Accessed + Protected + ReservedIM"),
                Permission::Active => throw_ub!("An actively protected pointer with Active permission becomes Disabled"),
                Permission::Frozen => throw_ub!("An actively protected pointer with Frozen permission becomes Disabled"),
                Permission::Reserved { .. } => throw_ub!("An actively protected pointer with Reserved permission becomes Disabled"),
            }
        }
        self.permission = permission;

>         self.accessed = self.accessed.transition(node_relation);

Reviewer: For `accessed` it is in fact crucial to do this before the UB check, I think? If this `accessed` is "no" before this access but becomes "yes" now and the state becomes (or already is) "Disabled", that should be UB!

I don't know if that state can actually be reached though.

        node_relation: NodeRelation,
        protected: bool,
    ) -> Result {
        let old_perm = self.permission;
        self.permission = old_perm.transition(access_kind, node_relation, protected)?;
        self.accessed = self.accessed.transition(node_relation);

>         if self.permission == Permission::Disabled && protected && self.accessed == Accessed::Yes {

Reviewer: ```suggestion
        // Protected nodes may never transition to "Disabled, Accessed". That is UB.
        if protected && self.permission == Permission::Disabled && self.accessed == Accessed::Yes {
```

            match old_perm {
                Permission::Disabled => panic!("Impossible state combination: Accessed + Protected + Disabled"),
                Permission::ReservedIM => panic!("Impossible state combination: Accessed + Protected + ReservedIM"),
                Permission::Active => throw_ub!("An actively protected pointer with Active permission becomes Disabled"),
                Permission::Frozen => throw_ub!("An actively protected pointer with Frozen permission becomes Disabled"),

>                 Permission::Reserved { .. } => throw_ub!("An actively protected pointer with Reserved permission becomes Disabled"),

Reviewer: ```suggestion
                Permission::Active => throw_ub!("A protected pointer with Active permission becomes Disabled"),
                Permission::Frozen => throw_ub!("A protected pointer with Frozen permission becomes Disabled"),
                Permission::Reserved { .. } => throw_ub!("A protected pointer with Reserved permission becomes Disabled"),
```

        self.permission = old_perm.transition(access_kind, node_relation, protected)?;
        self.accessed = self.accessed.transition(node_relation);

        // Protected nodes may never transition to "Disabled, Accessed". That is UB.
        if self.permission == Permission::Disabled && protected && self.accessed == Accessed::Yes {
            match old_perm {
                Permission::Disabled => panic!("Impossible state combination: Accessed + Protected + Disabled"),
                Permission::ReservedIM => panic!("Impossible state combination: Accessed + Protected + ReservedIM"),

>                 Permission::Active => throw_ub!("A protected pointer with Active permission becomes Disabled"),
>                 Permission::Frozen => throw_ub!("A protected pointer with Frozen permission becomes Disabled"),
>                 Permission::Reserved { .. } => throw_ub!("A protected pointer with Reserved permission becomes Disabled"),

Reviewer: Turns out we usually start these messages in lower case
```suggestion
                Permission::Active => throw_ub!("a protected pointer with Active permission becomes Disabled"),
                Permission::Frozen => throw_ub!("a protected pointer with Frozen permission becomes Disabled"),
                Permission::Reserved { .. } => throw_ub!("a protected pointer with Reserved permission becomes Disabled"),
```
